hybrid/per_func_comp_main.c

```diff
@@ -0,0 +1,257 @@
+    cheriobj_2_codecap = cheri_seal(
+        codecap_create((void (*)(void))(&sandbox_2_func), cheriobj_2_code_end), sealcap_2);
+
+    assert((size_t) &sandbox_1_func != (size_t) &sandbox_2_func);
+
+    //- Create cheri_object_2 *data* cap -//
+    void *cheriobj_2_data_base = (void *) ((size_t) apprx_data_UB - 1000); // approx
+    cheriobj_2_datacap = cheri_seal(datacap_create(cheriobj_2_data_base, apprx_data_UB), sealcap_2);
+
+    // Check
+    assert(cheri_is_sealed(cheriobj_1_codecap));
+    assert(cheri_is_sealed(cheriobj_1_datacap));
+    assert(cheri_is_sealed(cheriobj_2_codecap));
+    assert(cheri_is_sealed(cheriobj_2_datacap));
+
+    printf("____global_sealcap__________\n");
+    print_cap(global_sealcap);
+
+    printf(">>>> cheri_object_1__________________________\n");
+    printf("Sealcap_1_user_defined_tysz : %04lx\n", sealcap_size_1);
+    printf("___sealcap_1_info____\n");
+    print_cap(sealcap_1);
+    printf("___code_cap_1_sealed_info____\n");
+    print_cap(cheriobj_1_codecap);
+    printf("___data_cap_1_sealed_info____\n");
+    print_cap(cheriobj_1_datacap);
+
+    printf(">>>> cheri_object_2__________________________\n");
+
+    printf("Sealcap_2_user_defined_tysz : %04lx\n", sealcap_size_2);
+    printf("___sealcap_2_info____\n");
+    print_cap(sealcap_2);
+    printf("___code_cap_2_sealed_info____\n");
+    print_cap(cheriobj_2_codecap);
+    printf("___data_cap_2_sealed_info____\n");
+    print_cap(cheriobj_2_datacap);
+
+    printf(" ** code_1_type: %04x\n", cheri_type_get(cheriobj_1_codecap));
+    printf(" ** data_1_type: %04x\n", cheri_type_get(cheriobj_1_datacap));
+    printf(" ** code_2_type: %04x\n", cheri_type_get(cheriobj_2_codecap));
+    printf(" ** data_2_type: %04x\n\n", cheri_type_get(cheriobj_2_datacap));
+
+    // User-defined cheri_object's otype should be within valid range
+    assert(CHERI_OTYPE_USER_MIN <= cheri_type_get(cheriobj_1_codecap) &&
+           cheri_type_get(cheriobj_1_codecap) <= CHERI_OTYPE_USER_MAX);
+    assert(CHERI_OTYPE_USER_MIN <= cheri_type_get(cheriobj_2_codecap) &&
+           cheri_type_get(cheriobj_2_codecap) <= CHERI_OTYPE_USER_MAX);
+
+    printf(" ** Success: all code/data capability types are within valid range ** \n");
+
+    // Cheri_object_1 and 2's otypes should differ
+    assert(cheri_type_get(cheriobj_1_codecap) != cheri_type_get(cheriobj_2_codecap));
+    assert(cheri_type_get(cheriobj_1_datacap) != cheri_type_get(cheriobj_2_datacap));
+
+    printf(" ** Success: object 1 and 2's types differ ** \n");
+
+    // Each cheri_object's code and data address ranges are set the same,
+    // but perms for code and data differ.
+    // TODO: Test obj cap invoke
+    assert(cheri_perms_get(cheriobj_1_codecap) != cheri_perms_get(cheriobj_1_datacap));
+    assert(cheri_perms_get(cheriobj_2_codecap) != cheri_perms_get(cheriobj_2_datacap));
+    printf(" ** Success: code and data capability perms differ ** \n");
+
+    // TODO: cheri_invoke
+    sandbox_1_func();
```

Comment:
I guess this program doesn't compile at the moment because these function calls reference commented code?

Reply:
There are non-commented versions of these two functions [1]. But granted, I think some commented code cleaning is overall in order.

[1] https://github.com/capablevms/cheri-examples/pull/85/files#diff-f66e65b6692b851506e5182a9bfd264a08d4cab560e6ad190ac61beff5ae7283R127-R135